### ISSUE TITLE
sphinx.search.IndexBuilder.load() followed by prune() appears to be broken

### ISSUE THREAD

**Issue by nnemkin** [sphinx-doc/sphinx#3349]
`sphinx.search.IndexBuilder.load()` [here](https://github.com/sphinx-doc/sphinx/blob/master/sphinx/search/__init__.py#L296-L297) uses _filenames_ for `self._titles` and `self._mapping`, while it should be using _docnames_ instead:

```
index2fn = frozen['docnames']
```

`IndexBuilder._titles` is documented to use docnames and all of the code is obviously assuming docnames to be there.

I've discovered this problem when the following code failed in my custom builder:
```python
# self is the Builder, self.indexer is sphinx.search.IndexBuilder
self.indexer.feed(...)  # index documents, self.indexer._titles has docnames
self.indexer.dump(...)  # save searchindex.js
self.indexer.load(...)  # load searchindex.js, self.indexer._titles has filenames!
self.indexer.prune(self.env.all_docs)  # all docs are pruned erroneously
                                       # because IndexBuilder can't find docnames in _titles
```

**Comment by tk0miya** [comment]
+1
As a quick look, it seems your proposal is right.

Note: We should add testcase for the `IndexBuilder.load()` (and other methods).

**Comment by tk0miya** [comment]
Fixed at #3434.
Thank you for reporting!
